File: packages/core/src/validate.ts
import type { ErrorCode } from "./wire/errors.js";
import type { Op } from "./wire/ops.js";
import type { Attribute, AttributeType, Kind, WireSchema } from "./wire/schema.js";

/**
 * The verdict of {@link validate}: accepted, or rejected with exactly one
 * closed-enum code and a precise message. Rejection is all-or-nothing — the
 * caller applies nothing.
 */
export type ValidationResult = { ok: true } | { ok: false; code: ErrorCode; message: string };

const accept: ValidationResult = { ok: true };
const reject = (code: ErrorCode, message: string): ValidationResult => ({
  ok: false,
  code,
  message,
});

/** Check one scalar value against a declared attribute type. Returns the
 * failing code, or null if it matches. `element` values are ids that must be
 * live (already present or minted earlier in the batch). */
function checkScalar(
  value: unknown,
  type: AttributeType,
  liveIds: ReadonlySet<string>,
): ErrorCode | null {
  switch (type) {
    case "string":
      return typeof value === "string" ? null : "wrong-type";
    case "number":
      return typeof value === "number" ? null : "wrong-type";
    case "boolean":
      return typeof value === "boolean" ? null : "wrong-type";
    case "json":
      return null;
    case "element":
      if (typeof value !== "string") return "wrong-type";
      return liveIds.has(value) ? null : "bad-ref";
  }
}

/** Check an attribute's value, honouring `many` (a list of the base type). */
function checkValue(
  value: unknown,
  attr: Attribute,
  liveIds: ReadonlySet<string>,
): ErrorCode | null {
  if (attr.many) {
    if (!Array.isArray(value)) return "wrong-type";
    for (const item of value) {
      const code = checkScalar(item, attr.type, liveIds);
      if (code) return code;
    }
    return null;
  }
  return checkScalar(value, attr.type, liveIds);
}

/** Precise typed-failure message, carrying the attribute's description. */
function typedMessage(attr: Attribute, code: ErrorCode): string {
  const expected = attr.many ? `${attr.type}[]` : attr.type;
  if (code === "bad-ref") {
    return `attribute "${attr.name}" (${attr.description}) references an unknown element`;
  }
  return `attribute "${attr.name}" (${attr.description}) has the wrong type; expected ${expected}`;
}

/**
 * Stateless, typed, all-or-nothing validation of a flat ordered ops list
 * against a host schema. Returns the first failure, or acceptance.
 *
 * @param wireSchema  the board's declared host schema.
 * @param ops         the `apply` ops, validated in order.
 * @param existingIds ids already present on the board (empty for a fresh
 *   board). Ops may reference ids minted earlier in the same batch.
 *
 * Within-batch minting and deletion are tracked. Undeclared `data` fields pass
 * through untouched. Updates to elements minted in this batch are type-checked
 * against their kind (partial merge — absent attributes are unchanged);
 * updates to pre-existing ids are existence-checked only, since the caller
 * supplies ids without kinds.
 * ponytail: update type-check covers in-batch elements only; if A3 needs it for
 * pre-existing elements, widen `existingIds` to carry kinds.
 */
export function validate(
  wireSchema: WireSchema,
  ops: readonly Op[],
  existingIds: ReadonlySet<string>,
): ValidationResult {
  const kinds = new Map<string, Kind>(wireSchema.kinds.map((k) => [k.id, k]));
  const live = new Set(existingIds);
  const mintedKinds = new Map<string, string>();

  for (const op of ops) {
    if (op.op === "create") {
      const { id, kind, data } = op.element;
      const declared = kinds.get(kind);
      if (!declared) return reject("unknown-kind", `unknown kind: ${kind}`);
      if (live.has(id)) return reject("duplicate-id", `element id already exists: ${id}`);
      for (const attr of declared.attributes) {
        if (!Object.hasOwn(data, attr.name)) {
          if (attr.required) {
            return reject(
              "missing-required",
              `missing required attribute "${attr.name}": ${attr.description}`,
            );
          }
          continue;
        }
        const code = checkValue(data[attr.name], attr, live);
        if (code) return reject(code, typedMessage(attr, code));
      }
      live.add(id);
      mintedKinds.set(id, kind);
    } else if (op.op === "update") {
      if (!live.has(op.id)) return reject("unknown-element", `unknown element: ${op.id}`);
      const kindId = mintedKinds.get(op.id);
      const declared = kindId === undefined ? undefined : kinds.get(kindId);
      if (declared) {
        for (const attr of declared.attributes) {
          if (!Object.hasOwn(op.data, attr.name)) continue;
          const code = checkValue(op.data[attr.name], attr, live);
          if (code) return reject(code, typedMessage(attr, code));
        }
      }
    } else {
      if (!live.has(op.id)) return reject("unknown-element", `unknown element: ${op.id}`);
      live.delete(op.id);
      mintedKinds.delete(op.id);
    }
  }

  return accept;
}

File: packages/python/src/wboard/core/validate.py
"""Stateless, typed, all-or-nothing validation — the twin of ``@wboard/core``'s
``validate()``. Returns the first failure, or acceptance. The verdict must match
the TS side on every corpus fixture.
"""

import math
from collections.abc import Mapping, Sequence

from .errors import ErrorCode
from .wire import (
    ApplyAccepted,
    ApplyRejected,
    ApplyResponse,
    Attribute,
    AttributeType,
    CreateOp,
    DeleteOp,
    Op,
    UpdateOp,
    WireSchema,
)

_ACCEPT = ApplyAccepted(ok=True)


def _reject(code: ErrorCode, message: str) -> ApplyRejected:
    return ApplyRejected(ok=False, code=code, message=message)


def _is_number(value: object) -> bool:
    # A JSON number is finite and never a bool. ``isinstance(True, int)`` is a
    # Python-only trap the TS side never faces, so exclude bool explicitly.
    if isinstance(value, bool):
        return False
    if isinstance(value, int):
        # JS turns a huge JSON number into Infinity, which Number.isFinite
        # rejects; a Python int of any magnitude would pass. Match JS: a value
        # that overflows float64 is not a finite number.
        try:
            return math.isfinite(float(value))
        except OverflowError:
            return False
    if isinstance(value, float):
        return math.isfinite(value)
    return False


def _check_scalar(
    value: object, attr_type: AttributeType, live_ids: frozenset[str] | set[str]
) -> ErrorCode | None:
    """One scalar value against a declared type. ``element`` values are ids that
    must be live (already present or minted earlier in the batch)."""
    if attr_type == "string":
        return None if isinstance(value, str) else "wrong-type"
    if attr_type == "number":
        return None if _is_number(value) else "wrong-type"
    if attr_type == "boolean":
        return None if isinstance(value, bool) else "wrong-type"
    if attr_type == "json":
        return None
    # element
    if not isinstance(value, str):
        return "wrong-type"
    return None if value in live_ids else "bad-ref"


def _check_value(value: object, attr: Attribute, live_ids: set[str]) -> ErrorCode | None:
    """An attribute's value, honouring ``many`` (a list of the base type)."""
    if attr.many:
        if not isinstance(value, list):
            return "wrong-type"
        for item in value:
            code = _check_scalar(item, attr.type, live_ids)
            if code:
                return code
        return None
    return _check_scalar(value, attr.type, live_ids)


def _typed_message(attr: Attribute, code: ErrorCode) -> str:
    """Precise typed-failure message, carrying the attribute's description."""
    expected = f"{attr.type}[]" if attr.many else attr.type
    if code == "bad-ref":
        return f'attribute "{attr.name}" ({attr.description}) references an unknown element'
    return f'attribute "{attr.name}" ({attr.description}) has the wrong type; expected {expected}'


def validate(
    wire_schema: WireSchema,
    ops: Sequence[Op],
    existing: Mapping[str, str],
) -> ApplyResponse:
    """Validate a flat ordered ops list against a host schema.

    ``existing`` maps each already-present id to its kind (empty for a fresh
    board); the kind lets updates to pre-existing elements be type-checked.
    Within-batch minting and deletion are tracked. Undeclared ``data`` fields
    pass through. A create reusing any id live OR minted earlier in this batch
    (even if since deleted) is a ``duplicate-id``.
    """
    kinds = {k.id: k for k in wire_schema.kinds}
    live: set[str] = set(existing.keys())
    kind_of: dict[str, str] = dict(existing)
    ever_minted: set[str] = set()

    for op in ops:
        if isinstance(op, CreateOp):
            element = op.element
            declared = kinds.get(element.kind)
            if declared is None:
                return _reject("unknown-kind", f"unknown kind: {element.kind}")
            if element.id in live or element.id in ever_minted:
                return _reject("duplicate-id", f"element id already exists: {element.id}")
            for attr in declared.attributes:
                if attr.name not in element.data:
                    if attr.required:
                        return _reject(
                            "missing-required",
                            f'missing required attribute "{attr.name}": {attr.description}',
                        )
                    continue
                code = _check_value(element.data[attr.name], attr, live)
                if code:
                    return _reject(code, _typed_message(attr, code))
            live.add(element.id)
            ever_minted.add(element.id)
            kind_of[element.id] = element.kind
        elif isinstance(op, UpdateOp):
            if op.id not in live:
                return _reject("unknown-element", f"unknown element: {op.id}")
            kind_id = kind_of.get(op.id)
            declared = None if kind_id is None else kinds.get(kind_id)
            # A live id whose kind the schema does not declare cannot be
            # type-checked — fail closed as unknown-kind, never skip the checks.
            if declared is None:
                return _reject("unknown-kind", f"unknown kind: {kind_id or '(unmapped)'}")
            for attr in declared.attributes:
                if attr.name not in op.data:
                    continue
                code = _check_value(op.data[attr.name], attr, live)
                if code:
                    return _reject(code, _typed_message(attr, code))
        elif isinstance(op, DeleteOp):
            if op.id not in live:
                return _reject("unknown-element", f"unknown element: {op.id}")
            live.discard(op.id)

    return _ACCEPT
